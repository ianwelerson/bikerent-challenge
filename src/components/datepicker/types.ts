export interface SelectedDates {
  start: null | string
  end: null | string
}
