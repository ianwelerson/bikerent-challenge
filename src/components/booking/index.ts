export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
export { default as BookingOverview } from './BookingOverview.vue'
export { default as BookingSuccess } from './BookingSuccess.vue'
