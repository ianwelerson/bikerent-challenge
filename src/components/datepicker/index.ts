export { default as Datepicker } from './DatepickerInput.vue'
