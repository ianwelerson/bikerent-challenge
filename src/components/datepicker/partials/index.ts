export { default as MonthdayButton } from './MonthdayButton.vue'
