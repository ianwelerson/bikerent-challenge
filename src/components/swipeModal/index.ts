export { default as SwipeModal } from './SwipeModal.vue'
