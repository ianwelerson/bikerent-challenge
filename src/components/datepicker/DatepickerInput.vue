<script lang="ts">
import { defineComponent } from 'vue'
import type { PropType } from 'vue'

import type { SelectedDates } from './types'

import { generateDatepickerDays } from '@/core/helpers/date'

import { Icon } from '@/components/icon'
import { MonthdayButton } from '@/components/datepicker/partials'

export default defineComponent({
  name: 'DatepickerInput',
  components: {
    Icon,
    MonthdayButton
  },
  props: {
    modelValue: {
      type: Object as PropType<SelectedDates>,
      required: true
    },
    allowSingleDay: {
      type: Boolean,
      default: true
    },
    allowPastDate: {
      type: Boolean,
      default: false
    }
  },
  emits: ['update:modelValue'],
  data: () => ({
    month: new Date().getMonth() + 1,
    year: new Date().getFullYear(),
    selected: {
      start: null,
      end: null
    } as SelectedDates
  }),
  computed: {
    days() {
      return generateDatepickerDays(this.year, this.month, 35)
    },
    monthName() {
      return new Date(this.year, this.month, 0).toLocaleString('en-US', { month: 'long' })
    },
    disablePrevAction(): boolean {
      if (this.allowPastDate) {
        return false
      }

      return this.month <= new Date().getMonth() + 1
    }
  },
  watch: {
    selected: {
      handler(newDates) {
        if (!!newDates.start && !!newDates.end) {
          this.$emit('update:modelValue', Object.assign({}, newDates))
        }
      },
      deep: true
    }
  },
  mounted() {
    this.selected = this.modelValue
  },
  methods: {
    navigate(direction: 'PREV' | 'NEXT') {
      if (direction === 'PREV') {
        this.year = this.month === 1 ? this.year - 1 : this.year
        this.month = this.month > 1 ? this.month - 1 : 12
        return
      }

      this.year = this.month === 12 ? this.year + 1 : this.year
      this.month = this.month < 12 ? this.month + 1 : 1
    },
    setDate(date: string) {
      // Set start date
      if (!this.selected.start || !!this.selected.end) {
        this.selected.start = date
        this.selected.end = null
        return
      }

      // Switch dates if the end date is before the start
      if (date < this.selected.start) {
        this.selected.end = this.selected.start
        this.selected.start = date
        return
      }

      // Single day
      if (!this.allowSingleDay && date === this.selected.start) {
        return
      }

      this.selected.end = date
    }
  }
})
</script>

<template>
  <div class="datepicker">
    <div class="datepicker__wrapper">
      <div class="datepicker__header">
        <div class="current-month">
          <h3 class="current-month__label">{{ monthName }}</h3>
          <p class="current-month__year">{{ year }}</p>
        </div>
        <div class="navigation">
          <button
            :class="['navigation__button', { 'navigation__button--disabled': disablePrevAction }]"
            :disabled="disablePrevAction"
            @click="navigate('PREV')"
          >
            <div class="navigation__icon">
              <icon type="solid" size="2xl">chevron-left</icon>
            </div>
          </button>
          <button class="navigation__button" @click="navigate('NEXT')">
            <div class="navigation__icon">
              <icon type="solid" size="2xl">chevron-right</icon>
            </div>
          </button>
        </div>
      </div>
      <div class="datepicker__calendar">
        <div class="calendar">
          <div class="calendar__header">
            <div class="weekday">
              <p class="weekday__label">Su</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">Mo</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">Tu</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">We</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">Th</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">Fr</p>
            </div>
            <div class="weekday">
              <p class="weekday__label">Sa</p>
            </div>
          </div>
          <div class="calendar__days">
            <monthday-button
              v-for="day in days"
              :key="day"
              :current-month="month"
              :day="day"
              :selected="selected"
              @select="setDate"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<style scoped lang="scss">
.datepicker {
  background-color: get-theme-color('primary');
  color: get-theme-color('white');
  border-radius: $border-radius-xl;

  &__wrapper {
    @apply .pt-6, .px-6, .pb-10;
  }

  &__header {
    display: flex;
    justify-content: space-between;
  }

  &__calendar {
    @apply .mt-10;
  }
}

.current-month {
  &__label {
    @apply .font-extrabold, .text-2xl;
  }

  &__year {
    @apply .m-0, .font-semibold, .text-base;

    opacity: 0.5;
  }
}

.navigation {
  &__button {
    @apply .p-3;

    outline: none;
    background: transparent;
    border: 1px solid get-theme-color('white');
    border-radius: $border-radius-lg;
    cursor: pointer;

    &--disabled {
      opacity: 0.5;
      cursor: not-allowed;
    }
  }

  &__button:first-child {
    @apply .mr-2;
  }

  &__icon {
    height: 24px;
    width: 24px;
    color: get-theme-color('white');
  }
}

.calendar {
  &__header {
    display: flex;
  }

  &__days {
    @apply .mt-5;

    margin-left: -10px;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }
}

.weekday {
  flex: 1;

  &__label {
    @apply .m-0, .font-bold, .text-base;

    opacity: 0.5;
  }
}
</style>
