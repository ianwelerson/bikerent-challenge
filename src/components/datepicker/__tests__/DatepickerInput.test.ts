import { describe, expect, it, vi } from 'vitest'
import { nextTick } from 'vue'

import { DatepickerInput, type SelectedDates } from '@/components/datepicker'
import { mount } from '@vue/test-utils'

const FAKE_START_DATE = '2024-01-16'
const FAKE_END_DATE = '2024-01-20'

vi.useFakeTimers().setSystemTime(new Date(FAKE_START_DATE))

const selectedDate: SelectedDates = {
  start: FAKE_START_DATE,
  end: FAKE_END_DATE
}

vi.mock('@/composables/screenSize', () => ({
  useScreenSize: vi
    .fn()
    .mockReturnValueOnce({
      isMobile: true
    })
    .mockReturnValueOnce({
      isMobile: false
    })
    .mockReturnValueOnce({
      isMobile: false
    })
    .mockReturnValueOnce({
      isMobile: false
    })
    .mockReturnValueOnce({
      isMobile: false
    })
    .mockReturnValue({
      isMobile: true
    })
}))

describe('DatepickerInput', () => {
  it('renders correctly', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have the default dates selected', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    expect(wrapper.find('.monthday--start').text()).equals(FAKE_START_DATE.split('-')[2])
    expect(wrapper.findAll('.monthday--between')).toHaveLength(3)
    expect(wrapper.find('.monthday--end').text()).equals(FAKE_END_DATE.split('-')[2])
  })

  it('should have disabled dates based on the current day', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    expect(wrapper.findAll('.monthday--disabled')).toHaveLength(15)
  })

  it('should be able to navigate to the next month', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    expect(wrapper.find('.current-month__label').text()).toBe('January')

    await wrapper.find('.navigation__button:not(.navigation__button--disabled)').trigger('click')

    expect(wrapper.find('.current-month__label').text()).toBe('February')
  })

  it('should be able to select dates and receive the event', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    const availableDates = wrapper.findAll(
      '.monthday:not(.monthday--diff-month,.monthday--between,.monthday--start,.monthday--end,.monthday--disabled)'
    )
    const startButton = availableDates[availableDates.length - 4]
    const endButton = availableDates[availableDates.length - 1]

    await startButton.trigger('click')
    await endButton.trigger('click')

    expect(wrapper.find('.monthday--start').text()).equals('28')
    expect(wrapper.find('.monthday--end').text()).equals('31')

    expect((wrapper.emitted()['update:modelValue'][1] as SelectedDates[])[0]).toMatchObject({
      start: '2024-01-28',
      end: '2024-01-31'
    })
  })

  it('should be able to select dates and trigger the emit when is mobile', async () => {
    const wrapper = mount(DatepickerInput, { props: { modelValue: { ...selectedDate } } })

    await nextTick()

    const availableDates = wrapper.findAll(
      '.monthday:not(.monthday--diff-month,.monthday--between,.monthday--start,.monthday--end,.monthday--disabled)'
    )
    const startButton = availableDates[availableDates.length - 4]
    const endButton = availableDates[availableDates.length - 1]

    await startButton.trigger('click')
    await endButton.trigger('click')

    await wrapper.find('button.button--secondary').trigger('click')

    expect((wrapper.emitted()['update:modelValue'][0] as SelectedDates[])[0]).toMatchObject({
      start: '2024-01-28',
      end: '2024-01-31'
    })
  })
})
