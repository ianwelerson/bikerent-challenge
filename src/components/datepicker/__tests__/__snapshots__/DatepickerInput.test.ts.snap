// Vitest Snapshot v1

exports[`DatepickerInput > renders correctly 1`] = `
"<div class=\\"datepicker\\" data-v-b40e904a=\\"\\">
  <div class=\\"datepicker__wrapper\\" data-v-b40e904a=\\"\\">
    <div class=\\"datepicker__header\\" data-v-b40e904a=\\"\\">
      <div class=\\"current-month\\" data-v-b40e904a=\\"\\">
        <h3 class=\\"current-month__label\\" data-v-b40e904a=\\"\\">January</h3>
        <p class=\\"current-month__year\\" data-v-b40e904a=\\"\\">2024</p>
      </div>
      <div class=\\"navigation\\" data-v-b40e904a=\\"\\"><button class=\\"navigation__button navigation__button--disabled\\" disabled=\\"\\" data-v-b40e904a=\\"\\">
          <div class=\\"navigation__icon\\" data-v-b40e904a=\\"\\">
            <div class=\\"icon text-2xl\\" data-v-6ea13f32=\\"\\" data-v-b40e904a=\\"\\"><em class=\\"fa-solid fa-chevron-left\\" data-v-6ea13f32=\\"\\"></em></div>
          </div>
        </button><button class=\\"navigation__button\\" data-v-b40e904a=\\"\\">
          <div class=\\"navigation__icon\\" data-v-b40e904a=\\"\\">
            <div class=\\"icon text-2xl\\" data-v-6ea13f32=\\"\\" data-v-b40e904a=\\"\\"><em class=\\"fa-solid fa-chevron-right\\" data-v-6ea13f32=\\"\\"></em></div>
          </div>
        </button></div>
    </div>
    <div class=\\"datepicker__calendar\\" data-v-b40e904a=\\"\\">
      <div class=\\"calendar\\" data-v-b40e904a=\\"\\">
        <div class=\\"calendar__header\\" data-v-b40e904a=\\"\\">
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Su</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Mo</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Tu</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">We</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Th</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Fr</p>
          </div>
          <div class=\\"weekday\\" data-v-b40e904a=\\"\\">
            <p class=\\"weekday__label\\" data-v-b40e904a=\\"\\">Sa</p>
          </div>
        </div>
        <div class=\\"calendar__days\\" data-v-b40e904a=\\"\\"><button class=\\"monthday monthday--diff-month monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">31</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">01</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">02</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">03</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">04</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">05</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">06</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">07</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">08</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">09</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">10</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">11</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">12</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">13</span></button><button class=\\"monthday monthday--disabled\\" disabled=\\"\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">14</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">15</span></button><button class=\\"monthday monthday--start\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">16</span></button><button class=\\"monthday monthday--between\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">17</span></button><button class=\\"monthday monthday--between\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">18</span></button><button class=\\"monthday monthday--between\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">19</span></button><button class=\\"monthday monthday--end\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">20</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">21</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">22</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">23</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">24</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">25</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">26</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">27</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">28</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">29</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">30</span></button><button class=\\"monthday\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">31</span></button><button class=\\"monthday monthday--diff-month\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">01</span></button><button class=\\"monthday monthday--diff-month\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">02</span></button><button class=\\"monthday monthday--diff-month\\" data-v-06bf3248=\\"\\" data-v-b40e904a=\\"\\"><span class=\\"monthday__label\\" data-v-06bf3248=\\"\\">03</span></button></div>
      </div>
    </div>
    <div class=\\"mt-16\\" data-v-b40e904a=\\"\\"><button class=\\"button button--secondary w-full py-5\\" data-v-b40e904a=\\"\\">Select</button></div>
  </div>
</div>"
`;
